### ISSUE TITLE
Metric export fails

### ISSUE THREAD

**Issue by iNikem** [open-telemetry/opentelemetry-java-instrumentation#1111]
Current version prints the following exception during runtime:

```
[opentelemetry.auto.trace 2020-08-27 09:57:28:819 +0300] [grpc-default-executor-0] WARN io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter - Failed to export spans
io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service opentelemetry.proto.collector.metrics.v1.MetricsService
	at io.grpc.Status.asRuntimeException(Status.java:533)
	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:515)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:510)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:630)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:518)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:692)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:681)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by kubawach** [comment]
Happy to have a look :)

**Comment by malafeev** [comment]
reproduced using latest 0.8.0 version, but we wrapping agent in own.

**Comment by genebean** [comment]
I assume this is the same thing... my logs are being spammed with this coming from all 4 services I have using the java agent:

```plain
[opentelemetry.auto.trace 2020-09-30 06:04:51:291 -0700] [grpc-default-executor-0] WARN io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter - Failed to export metrics
io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service opentelemetry.proto.collector.metrics.v1.MetricsService
at io.grpc.Status.asRuntimeException(Status.java:533)
at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:515)
at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:510)
at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:630)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:518)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:692)
at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:681)
at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
```

This is a rather big issue for me due to the amount of spam it is adding to my system logs... hoping there is either a bug fix or some kind of workaround.

**Comment by benjaminhuo** [comment]
I can see similar error logs too, and I cannot see tracing data in jaeger, just wondering if this(no tracing data in jaeger) is related to this bug? @iNikem @kubawach @malafeev @genebean 

```bash
[opentelemetry.auto.trace 2020-10-07 10:18:11:238 +0000] [grpc-default-executor-5] WARN io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter - Failed to export metrics
io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service opentelemetry.proto.collector.metrics.v1.MetricsService
```

**Comment by genebean** [comment]
@benjaminhuo i am getting trace data so I imagine you’re having another issue.

**Comment by iNikem** [comment]
@benjaminhuo Indeed, this problem should NOT impact tracing data

**Comment by benjaminhuo** [comment]
Thanks for the quick response, guys.
I'll see why there is no tracing data.

**Comment by kubawach** [comment]
Cannot reproduce. Tried on two different laptops.

Testing performed:

- collector - latest (docker 3ebc9a107a85)
- agent:
  - default exporter (OLTP)
  - latest release (0.8.0)
  - newest (master, 0.9.0) built with Java 11
- JDK (runtime)
  - 1.8
  - 11
  - 14 

In all cases:
- no warnings in logs
- metrics exported successfully (logged o the collector side)
`agent logs metric exporter installation - TracerInstaller - Installed metric exporter: io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter`

Most probably RC of reported warnings was an issue in the collector that was fixed in the meantime. In case of a clear repro steps for the issue I’ll be more than happy to have another look into this.

FYI @iNikem - if you agree with above we may now close the issue

**Comment by iNikem** [comment]
Just reproduced:
1. checkout master, build agent as usual.
2. Get https://github.com/spring-petclinic/spring-petclinic-microservices application
3. Make the following compose file:

```
version: '2'

services:
  config-server:
    image: springcommunity/spring-petclinic-config-server
    container_name: config-server
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=config-server
    ports:
     - 8888:8888

  discovery-server:
    image: springcommunity/spring-petclinic-discovery-server
    container_name: discovery-server
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=discovery-server
    depends_on:
      - config-server
    entrypoint: ["./dockerize","-wait=tcp://config-server:8888","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
     - 8761:8761

  customers-service:
    image: springcommunity/spring-petclinic-customers-service
    container_name: customers-service
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=customers-service
    depends_on:
     - config-server
     - discovery-server
    entrypoint: ["./dockerize","-wait=tcp://discovery-server:8761","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
    - 8081:8081

  visits-service:
    image: springcommunity/spring-petclinic-visits-service
    container_name: visits-service
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=visits-service
    depends_on:
     - config-server
     - discovery-server
    entrypoint: ["./dockerize","-wait=tcp://discovery-server:8761","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
     - 8082:8082

  vets-service:
    image: springcommunity/spring-petclinic-vets-service
    container_name: vets-service
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=vets-service
    depends_on:
     - config-server
     - discovery-server
    entrypoint: ["./dockerize","-wait=tcp://discovery-server:8761","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
     - 8083:8083

  api-gateway:
    image: springcommunity/spring-petclinic-api-gateway
    container_name: api-gateway
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=api-gateway
    depends_on:
     - config-server
     - discovery-server
    entrypoint: ["./dockerize","-wait=tcp://discovery-server:8761","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
     - 8080:8080

  tracing-server:
    image: otel/opentelemetry-collector
    container_name: tracing-server
    mem_limit: 512M
    volumes:
    - ./otel:/otel
    command: --config /otel/otel.yaml
    ports:
     - 55679:55679
     - 55680:55680

  admin-server:
    image: springcommunity/spring-petclinic-admin-server
    container_name: admin-server
    mem_limit: 512M
    volumes:
      - ./otel:/otel
    environment:
      JAVA_TOOL_OPTIONS: -javaagent:/otel/opentelemetry-javaagent-all.jar
      OTEL_EXPORTER_OTLP_SPAN_ENDPOINT: tracing-server:55680
      OTEL_EXPORTER_OTLP_METRIC_ENDPOINT: tracing-server:55680
      OTEL_RESOURCE_ATTRIBUTES: service.name=admin-server
    depends_on:
     - config-server
     - discovery-server
    entrypoint: ["./dockerize","-wait=tcp://discovery-server:8761","-timeout=60s","--","java", "org.springframework.boot.loader.JarLauncher"]
    ports:
     - 9090:9090

  ## Grafana / Prometheus

  grafana-server:
    build: ./docker/grafana
    container_name: grafana-server
    mem_limit: 256M
    ports:
    - 3000:3000

  prometheus-server:
    build: ./docker/prometheus
    container_name: prometheus-server
    mem_limit: 256M
    ports:
    - 9091:9090
```

Give tracing server the following otel.yaml:
```
extensions:
  health_check:
  pprof:
    endpoint: 0.0.0.0:1777
  zpages:
    endpoint: 0.0.0.0:55679

receivers:
  zipkin:
  otlp:
    protocols:
      grpc:

processors:
  batch:
  queued_retry:

  attributes/newenvironment:
    actions:
      - key: environment
        value: nikem
        action: insert

exporters:
  logging:
    logLevel: debug

service:
  pipelines:
    traces:
      receivers: [zipkin, otlp]
      processors: [attributes/newenvironment, batch, queued_retry]
      exporters: [logging]

  extensions: [health_check, pprof, zpages]
```

Start everything: `docker-compose -f docker-compose-otel.yml`. Wait. Around every minute see

```
config-server        | [opentelemetry.auto.trace 2020-10-09 11:11:21:505 +0000] [grpc-default-executor-0] WARN io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter - Failed to export metrics
config-server        | io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service opentelemetry.proto.collector.metrics.v1.MetricsService
config-server        | 	at io.grpc.Status.asRuntimeException(Status.java:533)
config-server        | 	at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:515)
config-server        | 	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
config-server        | 	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
config-server        | 	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
config-server        | 	at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
config-server        | 	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
config-server        | 	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
config-server        | 	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
config-server        | 	at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
config-server        | 	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:510)
config-server        | 	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
config-server        | 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:630)
config-server        | 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:518)
config-server        | 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:692)
config-server        | 	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:681)
config-server        | 	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
config-server        | 	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
config-server        | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
config-server        | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
config-server        | 	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by kubawach** [comment]
@iNikem so, provided example managed to recreate the issue. This was however due to a metrics service not configured for the OTLP collector. After adding
`
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging]
`
to `otel.yaml` file, exceptions stopped and metrics started appearing in the collector log output:
`
tracing-server       | 2020-10-14T19:47:57.630Z DEBUG   loggingexporter/logging_exporter.go:406 ResourceMetrics #0
tracing-server       | Resource labels:
tracing-server       |      -> service.name: STRING(vets-service)
tracing-server       |      -> telemetry.auto.version: STRING(0.8.0)
tracing-server       |      -> telemetry.sdk.language: STRING(java)
tracing-server       |      -> telemetry.sdk.name: STRING(opentelemetry)
tracing-server       |      -> telemetry.sdk.version: STRING(0.8.0)
tracing-server       | InstrumentationLibraryMetrics #0
tracing-server       | InstrumentationLibrary io.opentelemetry.sdk.trace 
tracing-server       | Metric #0
tracing-server       | Descriptor:
tracing-server       |      -> Name: processedSpans
tracing-server       |      -> Description: The number of spans processed by the BatchSpanProcessor. [dropped=true if they were dropped due to high throughput]
tracing-server       |      -> Unit: 1
tracing-server       |      -> DataType: IntSum
tracing-server       |      -> IsMonotonic: true
tracing-server       |      -> AggregationTemporality: AGGREGATION_TEMPORALITY_CUMULATIVE
tracing-server       | IntDataPoints #0
tracing-server       | Data point labels:
tracing-server       |      -> spanProcessorType: BatchSpanProcessor
tracing-server       |      -> dropped: false
tracing-server       | StartTime: 1602704697737000000
tracing-server       | Timestamp: 1602704877592000000
tracing-server       | Value: 71
tracing-server       | ResourceMetrics #1
tracing-server       | Resource labels:
tracing-server       |      -> service.name: STRING(admin-server)
tracing-server       |      -> telemetry.auto.version: STRING(0.8.0)
tracing-server       |      -> telemetry.sdk.language: STRING(java)
tracing-server       |      -> telemetry.sdk.name: STRING(opentelemetry)
tracing-server       |      -> telemetry.sdk.version: STRING(0.8.0)
tracing-server       | InstrumentationLibraryMetrics #0
tracing-server       | InstrumentationLibrary io.opentelemetry.sdk.trace 
` 

In other words, the exception was simply gRPC way of saying that there is no metrics endpoint available as configured. I believe we can close the issue now.

**Comment by iNikem** [comment]
<sarcasm>Damn perfectly obviously error message</sarcasm>.

**Comment by kubawach** [comment]
> Damn perfectly obviously error message.

I blame gRPC ;)

**Comment by trask** [comment]
😭

**Comment by benjaminhuo** [comment]
the metrics part should be added to the pipeline
So this is a documentation issue or a default otel collector config issue?

**Comment by trask** [comment]
hey @benjaminhuo! I didn't follow your comment, can you explain more in detail?

**Comment by benjaminhuo** [comment]
@trask Sorry for the confusion, I mean will there be a fix for this issue?
A doc fix or a fix in the default otel collector config file?

**Comment by trask** [comment]
got it, we have #1393 to improve the error message from the Java agent side, I just added a comment there about also updating doc to mention this

**Comment by benjaminhuo** [comment]
Thanks for the update !

**Comment by RashmiRam** [comment]
> I assume this is the same thing... my logs are being spammed with this coming from all 4 services I have using the java agent:
> 
> ```
> [opentelemetry.auto.trace 2020-09-30 06:04:51:291 -0700] [grpc-default-executor-0] WARN io.opentelemetry.exporters.otlp.OtlpGrpcMetricExporter - Failed to export metrics
> io.grpc.StatusRuntimeException: UNIMPLEMENTED: unknown service opentelemetry.proto.collector.metrics.v1.MetricsService
> at io.grpc.Status.asRuntimeException(Status.java:533)
> at io.grpc.stub.ClientCalls$UnaryStreamToFuture.onClose(ClientCalls.java:515)
> at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
> at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
> at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
> at io.grpc.internal.CensusStatsModule$StatsClientInterceptor$1$1.onClose(CensusStatsModule.java:700)
> at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
> at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
> at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
> at io.grpc.internal.CensusTracingModule$TracingClientInterceptor$1$1.onClose(CensusTracingModule.java:399)
> at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:510)
> at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:66)
> at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.close(ClientCallImpl.java:630)
> at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl.access$700(ClientCallImpl.java:518)
> at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:692)
> at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:681)
> at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
> at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
> at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> at java.base/java.lang.Thread.run(Thread.java:834)
> ```
> 
> This is a rather big issue for me due to the amount of spam it is adding to my system logs... hoping there is either a bug fix or some kind of workaround.

@genebean 
To avoid flooding of these logs, I configured [`OTEL_IMR_EXPORT_INTERVAL`](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/IntervalMetricReader.java#L167) to be of higher value like `43200000`(12h). So that, the exporter is doing an export for every 12 hours. While doing export only, [the metrics were also collected](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/export/IntervalMetricReader.java#L192). So, I suppose no storing of metrics in between. We did this as stop gap to avoid so much error logs.
